File: client/src/modules/assetmanager/schemas/holding/valuation.schemas.ts
/**
 * Valuation Schemas
 *
 * Zod validation schemas for Valuation model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/holding_models.py
 * - Schema: /server/apps/assetmanager/schemas/holding_schemas/valuation_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/holding_subrouters/valuation_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// Valuation Schema (Full Representation)
// ==========================================

/**
 * Valuation schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class Valuation(BaseModel)
 */
export const ValuationSchema = z.object({
  id: z.number(),
  entity_id: z.number(),
  funding_round_id: z.number().nullable(),
  date: z.string(),
  valuation_value: z.number(),

  // Fund-specific fields (nullable — only fund-type entities use these)
  total_fund_units: z.number().nullable(),
  nav_per_share: z.number().nullable(),

  notes: z.string().nullable(),
  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(), // ISO datetime string from backend
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new valuation (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class ValuationCreate(BaseModel)
 */
export const CreateValuationSchema = z.object({
  entity_id: z.number(),
  valuation_value: z.number(),
  funding_round_id: z.number().nullable().optional(),
  date: z.string().nullable().optional(),

  // Fund-specific fields (nullable — only fund-type entities use these)
  total_fund_units: z.number().nullable().optional(),
  nav_per_share: z.number().nullable().optional(),

  notes: z.string().nullable().optional(),
});

/**
 * Schema for updating a valuation (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class ValuationUpdate(BaseModel)
 */
export const UpdateValuationSchema = z.object({
  entity_id: z.number().nullable().optional(),
  funding_round_id: z.number().nullable().optional(),
  date: z.string().nullable().optional(),
  valuation_value: z.number().nullable().optional(),

  // Fund-specific fields (nullable — only fund-type entities use these)
  total_fund_units: z.number().nullable().optional(),
  nav_per_share: z.number().nullable().optional(),

  notes: z.string().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type Valuation = z.infer<typeof ValuationSchema>;
export type CreateValuation = z.infer<typeof CreateValuationSchema>;
export type UpdateValuation = z.infer<typeof UpdateValuationSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single valuation
 * Backend equivalent: class ValuationResponse(BaseModel)
 */
export type ValuationResponse = {
  success: boolean;
  data?: Valuation;
  error?: string;
};

/**
 * Response containing multiple valuations
 * Backend equivalent: class ValuationsResponse(BaseModel)
 */
export type ValuationsResponse = {
  success: boolean;
  data?: Valuation[];
  error?: string;
};

File: server/apps/assetmanager/schemas/holding_schemas/valuation_schemas.py
"""
Valuation Schemas

Pydantic schemas for the Valuation model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from datetime import date as date_type

# ==========================================
# Valuation Schema (Full Representation)
# ==========================================

class Valuation(BaseModel):
    """Valuation schema - full representation"""
    id: int
    entity_id: int = Field(description="Associated entity ID")
    funding_round_id: int | None = Field(None, description="Associated funding round ID")
    date: date_type = Field(description="Valuation date")
    valuation_value: float = Field(description="Valuation value")

    # Fund-specific fields (nullable — only fund-type entities use these)
    total_fund_units: float | None = Field(None, description="Total units issued")
    nav_per_share: float | None = Field(None, description="valuation_value / total_fund_units")

    notes: str | None = Field(None, description="Additional notes")
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class ValuationCreate(BaseModel):
    """Schema for creating a new valuation"""
    entity_id: int = Field(description="Associated entity ID")
    valuation_value: float = Field(description="Valuation value")

    funding_round_id: int | None = Field(None, description="Associated funding round ID")
    date: date_type | None = Field(None, description="Valuation date")

    # Fund-specific fields (nullable — only fund-type entities use these)
    total_fund_units: float | None = Field(None, description="Total units issued")
    nav_per_share: float | None = Field(None, description="valuation_value / total_fund_units")

    notes: str | None = Field(None, description="Additional notes")

class ValuationUpdate(BaseModel):
    """Schema for updating a valuation"""
    entity_id: int | None = None
    funding_round_id: int | None = None
    date: date_type | None = None
    valuation_value: float | None = None

    # Fund-specific fields (nullable — only fund-type entities use these)
    total_fund_units: float | None = None
    nav_per_share: float | None = None

    notes: str | None = None

# ==========================================
# Response Types
# ==========================================

class ValuationResponse(BaseModel):
    """Response containing a single valuation"""
    success: bool
    data: Valuation | None = None
    error: str | None = None

class ValuationsResponse(BaseModel):
    """Response containing multiple valuations"""
    success: bool
    data: list[Valuation] | None = None
    error: str | None = None
